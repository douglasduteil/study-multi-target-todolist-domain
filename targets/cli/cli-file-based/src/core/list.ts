//

export interface FilterOptions {
  all: boolean;
  // XXX(douglasduteil): Move json out of here
  // json should be part of the "list" filter as it's format
  json: boolean;
  long: boolean;
}
