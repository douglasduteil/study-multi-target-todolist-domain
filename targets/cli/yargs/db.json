{
  "index": 0,
  "todos": []
}