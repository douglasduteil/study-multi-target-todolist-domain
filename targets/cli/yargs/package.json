{
  "name": "@todolist/target.yargs",
  "version": "0.0.0",
  "description": "",
  "private": true,
  "main": "index.js",
  "scripts": {
    "build": "ncc build ./src/index.ts -o dist",
    "lint": "tslint -p tsconfig.json -t stylish",
    "integration": "jest --config integration/jest.config.js",
    "test": "jest",
    "start": "ncc run ./src/index.ts",
    "watch": "yarn start --watch"
  },
  "keywords": [],
  "author": "Douglas Duteil <douglasduteil@gmail.com> (douglasduteil.github.io)",
  "license": "WTFPL",
  "dependencies": {
    "@todolist/core": "^1.0.0",
    "@todolist/databases.lowdb": "^1.0.0",
    "@todolist/ui.yargs": "^1.0.0",
    "lowdb": "^1.0.0"
  },
  "devDependencies": {
    "@types/jest": "^23.3.12",
    "@types/node": "^10.7.1",
    "@types/lowdb": "^1.0.6",
    "@zeit/ncc": "^0.8.1",
    "prettier": "^1.15.3",
    "ts-loader": "^5.3.1",
    "tslint": "^5.12.0",
    "typescript": "^3.2.2"
  }
}
