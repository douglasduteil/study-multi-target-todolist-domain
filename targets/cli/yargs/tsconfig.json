{
  "extends": "../../../tsconfig.json",
  "compilerOptions": {
    "resolveJsonModule": true
  },
  "files": ["src/index.ts"]
}
