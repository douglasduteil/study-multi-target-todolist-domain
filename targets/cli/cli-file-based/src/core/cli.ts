//

import yargs from "yargs";
import { version } from "../../package.json";

export function todolistCli(): yargs.Argv {
  return yargs
    .usage("Usage: $0 <command> [options]")
    .version(version)
    .demandCommand(
      1,
      "A command is required. Pass --help to see all available commands and options."
    )
    .recommendCommands()
    .strict()
    .alias("h", "help")
    .alias("v", "version");
}
