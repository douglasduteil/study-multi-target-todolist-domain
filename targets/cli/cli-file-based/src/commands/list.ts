//

import { pipe } from "ramda";
import { Arguments, Argv, Options } from "yargs";
import { ListOptions } from "../core/list";

//

export const aliases: ReadonlyArray<Aliases> = ["ls", "la", "ll"];
export const command = "list";
export const describe = "List todos";

// HACK(douglasduteil): CommandBuilder<any, any> ...
// I didn't find how to make tsc happy without any-s here...
// tslint:disable-next-line:typedef
export function builder<T>(yargs: Argv<T>): Argv<any> {
  return yargs.options(yargsOptions).group(Object.keys(yargsOptions), group);
}

export const handler: (args: ListArguments) => void = pipe(mapAliasOnOptions);

//

export type ListArguments = Arguments<{ [key in keyof ListOptions]: boolean }>;
export type Aliases = "ls" | "la" | "ll";

//

export const group = "List Options:";
export const ALIASES_DEFAULT_OPTIONS: {
  [key in Aliases]: Partial<ListOptions>
} = {
  la: { all: true },
  ll: { long: true },
  ls: {}
};

export const yargsOptions: { [key in keyof ListOptions]: Options } = {
  all: {
    alias: "a",
    default: false,
    describe: "Show done todos that are normally hidden",
    type: "boolean"
  },
  json: {
    default: false,
    describe: "Show information as a JSON array",
    type: "boolean"
  },
  long: {
    alias: "l",
    default: false,
    describe: "Show extended information",
    type: "boolean"
  }
};

//

export function mapAliasOnOptions(argv: ListArguments): ListArguments {
  const alias = argv._[0] as Aliases;
  const options = ALIASES_DEFAULT_OPTIONS[alias];
  return { ...argv, ...options };
}
