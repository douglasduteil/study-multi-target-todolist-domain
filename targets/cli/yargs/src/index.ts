//

import { filterTodosFn } from "@todolist/core";
import { Database } from "@todolist/databases.lowdb";
import { commands, main } from "@todolist/ui.yargs";
import LowdbFileSyncStorageAdapter from "lowdb/adapters/FileSync";
import { version } from "../package.json";

//

const lowDbDataSource = new Database(
  new LowdbFileSyncStorageAdapter("db.json")
);

const { AddCommand, ListCommand } = commands;

main({ version })
  .command(new AddCommand(lowDbDataSource))
  .command(new ListCommand(filterTodosFn, console, lowDbDataSource))
  .parse(process.argv.slice(2));
