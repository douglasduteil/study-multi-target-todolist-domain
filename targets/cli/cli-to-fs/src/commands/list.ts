//

import { ListCommand as InternalListCommand } from "@todolist.ui/yargs/lib/commands/list";
import { Todo } from "@todolist/core";
import { decorate, inject, injectable } from "inversify";
import filter from "ramda/es/filter";
import includes from "ramda/es/includes";
import map from "ramda/es/map";
import pickBy from "ramda/es/pickBy";
import pipe from "ramda/es/pipe";
import unless from "ramda/es/unless";

//

decorate(injectable(), InternalListCommand);

//

@injectable()
export class ListCommand extends InternalListCommand {
  constructor(@inject("DAL") private dal: any) {
    super();
  }

  public handler(args: any): void {
    console.log(this);
    const { all, long } = this.preHandler(args);
    const rawData = this.dal.todoDB.getTodos();

    const listFrom = dataFilterFn({ all, long });

    const data = listFrom(rawData);

    console.log(data);
  }
}

const filterUncompletedTodo = filter<Todo, "array">(todo => !todo.completed);
const filterMetaData = pickBy(
  includes<Array<keyof Todo>>(["completed", "id", "title"])
);

const dataFilterFn = (options: { all: boolean; long: boolean }) =>
  pipe(
    unless(() => options.all, filterUncompletedTodo),
    unless(() => options.long, map<Todo, Todo>(filterMetaData))
  );
