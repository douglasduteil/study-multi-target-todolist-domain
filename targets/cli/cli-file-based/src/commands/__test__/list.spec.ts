//

import yargs from "yargs";
import { FilterOptions } from "../../core/list";
import { builder, mapAliasOnOptions } from "../list";

//

const defaultArgv: FilterOptions = { all: false, json: false, long: false };

//

describe("builder", () => {
  it("should add list options to cli", () => {
    const log = jest.fn();
    builder(yargs.usage("foo list options")).showHelp(log as any);
    expect(log.mock.calls[0]).toMatchSnapshot();
  });
});

//

describe("mapAliasOnOptions", () => {
  test("should do nothing if no alias is given", () => {
    expect(
      mapAliasOnOptions({ $0: "foo", _: [], ...defaultArgv })
    ).toMatchSnapshot();
  });

  test("should do nothing if the alias is not found", () => {
    expect(
      mapAliasOnOptions({
        $0: "foo",
        _: ["oof"],
        ...defaultArgv
      })
    ).toMatchSnapshot();
  });

  test('"ls" do nothing', () => {
    expect(
      mapAliasOnOptions({
        $0: "foo",
        _: ["ls"],
        ...defaultArgv
      })
    ).toMatchSnapshot();
  });

  test('"la" should activate "all" option', () => {
    expect(
      mapAliasOnOptions({
        $0: "foo",
        _: ["la"],
        ...defaultArgv
      })
    ).toMatchSnapshot();
  });

  test('"ll" should activate "long" option', () => {
    expect(
      mapAliasOnOptions({
        $0: "foo",
        _: ["ll"],
        ...defaultArgv
      })
    ).toMatchSnapshot();
  });
});
