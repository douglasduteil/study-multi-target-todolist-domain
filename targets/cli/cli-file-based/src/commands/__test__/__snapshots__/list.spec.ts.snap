// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builder should add list options to cli 1`] = `
Array [
  "foo list <options>

List Options:
  --all, -a   Show done todos that are normally hidden[boolean] [default: false]
  --json      Show information as a JSON array        [boolean] [default: false]
  --long, -l  Show extended information               [boolean] [default: false]

Options:
  --help     Show help                                                 [boolean]
  --version  Show version number                                       [boolean]",
]
`;

exports[`mapAliasOnOptions "la" should activate "all" option 1`] = `
Object {
  "$0": "foo",
  "_": Array [
    "la",
  ],
  "all": true,
  "json": false,
  "long": false,
}
`;

exports[`mapAliasOnOptions "ll" should activate "long" option 1`] = `
Object {
  "$0": "foo",
  "_": Array [
    "ll",
  ],
  "all": false,
  "json": false,
  "long": true,
}
`;

exports[`mapAliasOnOptions "ls" do nothing 1`] = `
Object {
  "$0": "foo",
  "_": Array [
    "ls",
  ],
  "all": false,
  "json": false,
  "long": false,
}
`;

exports[`mapAliasOnOptions should do nothing if no alias is given 1`] = `
Object {
  "$0": "foo",
  "_": Array [],
  "all": false,
  "json": false,
  "long": false,
}
`;

exports[`mapAliasOnOptions should do nothing if the alias is not found 1`] = `
Object {
  "$0": "foo",
  "_": Array [
    "oof",
  ],
  "all": false,
  "json": false,
  "long": false,
}
`;
